tools/publisher/src/subcommand/fix_manifests.rs

```diff
@@ -0,0 +1,185 @@
+    metadata: &mut cargo_toml::Manifest,
+) -> Result<usize> {
+    let mut changed = fix_dep_set(versions, &mut metadata.dependencies)?;
+    changed += fix_dep_set(versions, &mut metadata.dev_dependencies)?;
+    changed += fix_dep_set(versions, &mut metadata.build_dependencies)?;
+    Ok(changed)
+}
+
+async fn fix_manifests(
+    fs: Fs,
+    versions: &BTreeMap<String, Version>,
+    manifests: &mut Vec<Manifest>,
+) -> Result<()> {
+    for manifest in manifests {
+        let changed = fix_dep_sets(versions, &mut manifest.metadata)?;
+        if changed > 0 {
+            let contents = toml::to_string(&manifest.metadata)
+                .with_context(|| format!("failed to serialize to toml for {:?}", manifest.path))?;
+            fs.write_file(&manifest.path, contents.as_bytes()).await?;
+            info!("Changed {} dependencies in {:?}.", changed, manifest.path);
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_fix_dep_sets() {
+        let manifest = br#"
+            [package]
+            name = "test"
+            version = "1.2.0-preview"
+
+            [build-dependencies]
+            build_something = "1.3"
+            local_build_something = { path = "../local_build_something" }
+
+            [dev-dependencies]
+            dev_something = "1.1"
+            local_dev_something = { path = "../local_dev_something" }
+
+            [dependencies]
+            something = "1.0"
+            local_something = { path = "../local_something" }
+        "#;
+        let metadata = cargo_toml::Manifest::from_slice(manifest).unwrap();
+        let mut manifest = Manifest {
+            path: "test".into(),
+            metadata,
+        };
+        let versions = vec![
+            ("local_build_something", "0.2.0"),
+            ("local_dev_something", "0.1.0"),
+            ("local_something", "1.1.3"),
+        ]
+        .into_iter()
+        .map(|e| (e.0.to_string(), Version::parse(e.1).unwrap()))
+        .collect();
+
+        fix_dep_sets(&versions, &mut manifest.metadata).expect("success");
+
+        let actual_deps = toml::Value::try_from(&manifest.metadata.dependencies).unwrap();
+        assert_eq!(
+            "\
+                something = \"1.0\"\n\
+                \n\
+                [local_something]\n\
+                features = []\n\
+                optional = false\n\
+                path = \"../local_something\"\n\
+                version = \"1.1.3\"\n\
+            ",
+            actual_deps.to_string()
+        );
+
+        let actual_dev_deps = toml::Value::try_from(&manifest.metadata.dev_dependencies).unwrap();
+        assert_eq!(
+            "\
+                dev_something = \"1.1\"\n\
+                \n\
+                [local_dev_something]\n\
+                features = []\n\
+                optional = false\n\
+                path = \"../local_dev_something\"\n\
+                version = \"0.1.0\"\n\
+            ",
```

Comment:
This may be more readable if defined with a [raw string literal](https://doc.rust-lang.org/rust-by-example/std/str.html#literals-and-escapes)